onnx/checker.cc

```diff
@@ -676,39 +676,85 @@ void check_graph(const GraphProto& graph, const CheckerContext& ctx, const Lexic
+    return;
+  }
+
+  const auto* schema_for_function_import =
+      ctx.get_schema_registry()->GetSchema(node.op_type(), func_opset_version, node.domain());
+
+  const auto* schema_for_model_import =
+      ctx.get_schema_registry()->GetSchema(node.op_type(), model_opset_version, node.domain());
+
+  if (schema_for_function_import && schema_for_model_import &&
+      schema_for_function_import->since_version() == schema_for_model_import->since_version()) {
+    // The since versions of schema for both opset imports match. This means they are compatible.
+    return;
+  }
+
+  fail_check(
+      "Opset import for domain " + node.domain() + " in function op " + node.op_type() +
+      "is not compatible with the version imported my model. FunctionOp imports version " +
+      ONNX_NAMESPACE::to_string(func_opset_version) + "whereas model imports version " +
+      ONNX_NAMESPACE::to_string(model_opset_version));
+}
+
 void check_function(const FunctionProto& function, const CheckerContext& ctx, const LexicalScopeContext& parent_lex) {
   enforce_non_empty_field(function, name);
-  enforce_has_field(function, since_version);
+  
+  if (ctx.get_ir_version() >= 0x00000008) {
+    enforce_has_field(function, domain);
+  }
 
   const auto& model_opset_imports = ctx.get_opset_imports();
   CheckerContext ctx_copy = ctx;
 
-  // A FunctionProto body (graph) may implicitly rely on the OperatorSet that
-  // this function belongs to or it can also explicitly rely on more OperatorSets
-  // specified in the opset_import field for FunctionProto. However the spec does not
-  // clarify how to treat inconsistencies between function level and model level OperatorSets imports.
-  // Right now we merge opset imports from function proto with the model level opset imports
-  // In case there is an inconsistency it is treated as an error.
-  // TODO: Clarify spec and revisit this check.
-  std::unordered_map<std::string, int> func_opset_imports{model_opset_imports};
+  std::unordered_map<std::string, int> func_opset_imports;
   for (auto& relied_opset : function.opset_import()) {
     auto& domain = relied_opset.domain();
     int version = static_cast<int>(relied_opset.version());
-    auto it = func_opset_imports.find(domain);
-    if (it == func_opset_imports.end()) {
-      func_opset_imports[domain] = version;
-    } else {
-      if (it->second != version) {
-        fail_check(
-            "ONNX models do not support multiple opset version imports for a domain. Function ",
-            function.name(),
-            " imports opset version ",
-            std::to_string(version),
-            " for domain ",
-            (domain.empty() ? "ai.onnx" : domain),
-            " where as the model imports opset version ",
-            std::to_string(it->second));
-      }
+    func_opset_imports[domain] = version;
+
+    auto it = model_opset_imports.find(domain);
+    if (it != model_opset_imports.end() && version > it->second) {
```

Comment:
Is this allowing the model to import opsets at any version less than the model imports for the same domain? (This seems different from the comment on the protobuf, but I haven't internalized all the logic yet, so I might be missing something....)

Reply:
I'm also not quite sure why this logic is here rather than in `check_opset_compatibility()`. Might be worth adding a comment to explain.

Reply:
Agree. It feels like this might be redundant, given check_opset_compatibility. If not, having the logic in one place seems better.